From 0a01fa92e29441eaa951280216736133219e5965 Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH 25/49] Sync mLastQueuedCrop with mLastQueuedSlot.

Because mLastQueuedCrop is not sync with mLastQueuedSlot at the
same mutex scope, it will cause getLastQueuedBuffer to compute
wrong transform matrix for output graphic buffer.
---
 libs/gui/BufferQueueProducer.cpp | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/libs/gui/BufferQueueProducer.cpp b/libs/gui/BufferQueueProducer.cpp
index 400ec8bf0..b373dbb8e 100644
--- a/libs/gui/BufferQueueProducer.cpp
+++ b/libs/gui/BufferQueueProducer.cpp
@@ -938,7 +938,6 @@ status_t BufferQueueProducer::queueBuffer(int slot,
 
         mCore->mBufferHasBeenQueued = true;
         mCore->mDequeueCondition.broadcast();
-        mCore->mLastQueuedSlot = slot;
 
         output->width = mCore->mDefaultWidth;
         output->height = mCore->mDefaultHeight;
@@ -987,9 +986,13 @@ status_t BufferQueueProducer::queueBuffer(int slot,
         connectedApi = mCore->mConnectedApi;
         lastQueuedFence = std::move(mLastQueueBufferFence);
 
-        mLastQueueBufferFence = std::move(acquireFence);
-        mLastQueuedCrop = item.mCrop;
-        mLastQueuedTransform = item.mTransform;
+        { // Autolock scope
+            Mutex::Autolock lock2(mCore->mMutex);
+            mLastQueueBufferFence = std::move(acquireFence);
+            mLastQueuedCrop = item.mCrop;
+            mLastQueuedTransform = item.mTransform;
+            mCore->mLastQueuedSlot = slot;
+        }
 
         ++mCurrentCallbackTicket;
         mCallbackCondition.broadcast();
-- 
2.11.0

